### PR TITLE
HIP-584: Add burn precompile

### LINKED ISSUE #5655: HIP-584: Provide implementation for mint and burn precompiles
### Problem

We need default implementations for precompiles related to mint and burn logic, so we should copy needed logic from hedera-services into hedera-mirror-web3.

### Solution

Copy precompile logic for the following precompiles:

- MintPrecompile
- BurnPrecompile

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:

Copied the support for BurnPrecompile

**Related issue(s)**:

Fixes #5655

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by 0xivanov — COMMENTED** [review]
Fix the code smells, you can use `@SuppressWarnings` where the code is copied from services

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    private final SyntheticTxnFactory syntheticTxnFactory;
+    private final BurnLogic burnLogic;
+
+    private BurnWrapper burnOp;
+
+    public BurnPrecompile(
+            final PrecompilePricingUtils pricingUtils,
+            final EncodingFacade encoder,
+            final SyntheticTxnFactory syntheticTxnFactory,
+            BurnLogic burnLogic) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java:79]: burnLogic could also be `final`

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Override
+    public Set<Integer> getFunctionSelectors() {
+        return Set.of(AbiConstants.ABI_ID_BURN_TOKEN, AbiConstants.ABI_ID_BURN_TOKEN_V2);
+    }
+
+    @Override
+    public Bytes getSuccessResultFor(final RunResult runResult) {
+        final var burnResult = (BurnResult) runResult;
+        return encoder.encodeBurnSuccess(burnResult.totalSupply());
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java:141]: Done 👍 I saw how you did it in MintPrecompile and did it the same way. I also refactored the code a bit and extracted the main logic into a private method leaving original methods as wrappers(so it won't bother the callers with details).
I also extracted the convertArrayToLong method in the MiscUtils class(not sure if that is the best place, if you have a better place in mind I can move it) so I can reuse it.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java (hunk 1 above)
Comment (vtronkov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java:79]: Done

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+ */
+public class BurnPrecompile extends AbstractWritePrecompile {
+
+    private static final List<Long> NO_SERIAL_NOS = Collections.emptyList();
+
+    private final EncodingFacade encoder;
+    private final SyntheticTxnFactory syntheticTxnFactory;
+    private final BurnLogic burnLogic;
+
+    private BurnWrapper burnOp;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java:74]: It's still stateful. Have to make stateless or create for every request.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java (hunk 3 above)
Comment (vtronkov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java]: Fixed 👍

**Comment by vtronkov** [comment]
> I have the feeling that our formatting is not applied in this PR. Would you mind checking that ?

Fixed

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            final var targetSerialNos = burnBody.getSerialNumbersList();
+            tokenModificationResult = burnLogic.burn(Id.fromGrpcToken(tokenId), 0, targetSerialNos, store);
+        } else {
+            tokenModificationResult =
+                    burnLogic.burn(Id.fromGrpcToken(tokenId), burnBody.getAmount(), NO_SERIAL_NOS, store);
+        }
+
+        final var modifiedToken = tokenModificationResult.token();
+        return new BurnResult(
+                modifiedToken.getTotalSupply(),
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/BurnPrecompile.java]: Fixed.
I also found that getMinimumFeeInTinybars was not implemented correctly so I fixed that.

**Review by steven-sheehy — APPROVED** [review]
LGTM, just need to merge

File: hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java [hunk 5]
```diff
@@ -79,7 +79,7 @@ public final class ParsingConstants {
     public static final TupleType intBoolTuple = TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
     public static final TupleType intAddressTuple = TupleType.parse("(int32,address)");
     public static final TupleType intPairTuple = TupleType.parse("(int32,int32)");
-    public static final TupleType burnReturnType = TupleType.parse("(int32,uint64)");
+    public static final TupleType burnReturnType = TupleType.parse("(int32,uint64,int64[])");
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/hapi/utils/contracts/ParsingConstants.java:82]: Having a look at services code, it seems for burn we return only the new total supply, so we should revert this change.

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java [hunk 6]
```diff
@@ -93,36 +94,36 @@ public Bytes encodeApproveNFT(final int responseCode) {
...
     }
 
-    public Bytes encodeMintFailure(@NonNull final ResponseCodeEnum status) {
-        return functionResultBuilder()
-                .forFunction(HAPI_MINT)
-                .withStatus(status.getNumber())
-                .withTotalSupply(0L)
-                .withSerialNumbers(NO_MINTED_SERIAL_NUMBERS)
-                .build();
+    public Bytes encodeBurnSuccess(final long totalSupply, final long[] serialNumbers) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java:100]: In services we encode only the `totalSupply`

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java [hunk 7]
```diff
@@ -93,36 +94,36 @@ public Bytes encodeApproveNFT(final int responseCode) {
...
+        return encodeFailure(HAPI_BURN, status);
+    }
+
+    private Bytes encodeFailure(final FunctionType functionType, final ResponseCodeEnum status) {
         return functionResultBuilder()
-                .forFunction(FunctionType.HAPI_BURN)
+                .forFunction(functionType)
                 .withStatus(status.getNumber())
                 .withTotalSupply(0L)
+                .withSerialNumbers(NO_SERIAL_NUMBERS)
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java:126]: We encode only the totalSupply in services

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**